### PR TITLE
[10.0][mig] account_invoice_merge

### PR DESCRIPTION
**Pull request by LoisRForgeFlow**
Task done in this migration:
* Replace openerp tags and import by odoo.
* Update README.
* Adapt header of the files to new template.
* Correct typo error in wizard view.
* Remove the use of workflow in the invoice.
* Squash commit history.

### CODE REVIEW

File: account_invoice_merge/models/account_invoice.py [hunk 1]
```diff
@@ -58,15 +47,16 @@ def _get_first_invoice_fields(self, invoice):
...
+            'payment_term_id': invoice.payment_term_id.id,
+            # 'period_id': invoice.period_id.id,
+            'invoice_line_ids': {},
             'partner_bank_id': invoice.partner_bank_id.id,
         }
 
     @api.multi
-    def do_merge(self, keep_references=True, date_invoice=False):
+    def do_merge(self, keep_references=True, date_invoice=False,
+                 remove_empty_invoice_lines=True):
```
Comment (adrienpeiffer) [inline account_invoice_merge/models/account_invoice.py:58]: Where this new option is used

**Review by adrienpeiffer — APPROVED** [review]
Just one question. Otherwise :+1: (Code review and functional testing)

File: account_invoice_merge/models/account_invoice.py (hunk 1 above)
Comment (lreficent) [inline account_invoice_merge/models/account_invoice.py]: That feature was added by @alexis-via

A fact is that Is default True so the corresponding part is always executed. But my guess is that it is added as an option for more customization possibilities in other modules.

Comment (LoisRForgeFlow) [reply account_invoice_merge/models/account_invoice.py:58]: That feature was added by @alexis-via

A fact is that it's by default 'True' so the corresponding part of code is always executed. But my guess is that it is added as an option for more customization possibilities in other modules.

**Review by chienandalu — APPROVED** [review]
tested functionality :+1:

**Comment by pedrobaeza** [comment]
I think tests for this module are very needed.

**Comment by JordiBForgeFlow** [comment]
@mreficent can you add unit tests to this one?

**Comment by LoisRForgeFlow** [comment]
Thank you @mreficent! This is ready to merge, right?

cc @pedrobaeza @jbeficent

**Comment by pedrobaeza** [comment]
Can you make a bit more asserts on the tests (like checking the number of lines in the resulting invoice, check that 2 invoices with different partner are not merged...)? If it's too much, I understand, and as a first test version can serve.

**Comment by sergio-teruel** [comment]
@lreficent This is not ready for merge... 
Can you rebase to review please??

**Comment by LoisRForgeFlow** [comment]
@pedrobaeza We'll try to improve a bit. :+1: 

@sergio-teruel well, all test are green, there are no conflicts, runbot is working... I don't see the problem to review it. Anyway, it was for sure rebased yesterday since @mreficent squashed the commits (for what you need to do a rebase).

**Comment by sergio-teruel** [comment]
@lreficent This module not works well. When I  merge a two invoices from one sale the sale line qty_invoiced field is set to zero, so i can invoice this sale order.
I try do a reviw but the code is not  in your diffs

**Comment by sergio-teruel** [comment]
@chienandalu Can test??

**Comment by carlosdauden** [comment]
@lreficent ¿Can you include this modification?:
https://github.com/OCA/account-invoicing/pull/262
Thanks

**Comment by MiquelRForgeFlow** [comment]
* Added the suggested tests by @pedrobaeza and some more.
* Solves sales object issue.
* Added OCA icon and summary info.

**Comment by sergio-teruel** [comment]
@mreficent You can not marge invoice lines thatn come in from different order_lines, if you do that the qty_invoiced field is wrong.
I attach images for better explanation:
I think that if invoice has a sale order you do not merge lines
seleccion_088
seleccion_090
seleccion_089

**Comment by MiquelRForgeFlow** [comment]
@sergio-teruel Ok, we will work in this issue.

**Comment by sergio-teruel** [comment]
@mreficent perfect... thanks!! :smile:

File: account_invoice_merge/models/account_invoice.py [hunk 2]
```diff
@@ -36,8 +24,9 @@ def _get_invoice_key_cols(self):
     @api.model
     def _get_invoice_line_key_cols(self):
         fields = [
-            'name', 'origin', 'discount', 'invoice_line_tax_id', 'price_unit',
+            'name', 'origin', 'discount', 'invoice_line_tax_ids', 'price_unit',
             'product_id', 'account_id', 'account_analytic_id',
+            'uom_id'
         ]
         for field in ['analytics_id']:
```
Comment (MiquelRForgeFlow) [inline account_invoice_merge/models/account_invoice.py]: The 'analytics_id' field was added by <code>account_analytic_plans</code> base module. But this base module was removed in v9. So this field is not anymore in versions 9.0 and 10.0

**Review by MiquelRForgeFlow — CHANGES_REQUESTED** [review]

**Review by MiquelRForgeFlow — APPROVED** [review]
LGTM! Please, review and then I will squash.
@lreficent @sergio-teruel @pedrobaeza

**Comment by LoisRForgeFlow** [comment]
LGTM :+1:

**Comment by sergio-teruel** [comment]
@lreficent Can we add the context key "is_merge" whan cancel the old invoices??

**Comment by rafaelbn** [comment]
Hi @mreficent last commit https://github.com/OCA/account-invoicing/pull/238/commits/9cc6a1c02fcd2787c7f7a5f1daa6b2bff77cc1ab introduce a travis error. Do you know why?

**Comment by MiquelRForgeFlow** [comment]
It's not from last commit. It's from #254 and their tests.

**Comment by rafaelbn** [comment]
@luismontalba could you please review https://github.com/OCA/account-invoicing/pull/238#issuecomment-303741400 thanks!

**Comment by pedrobaeza** [comment]
@mreficent I think the tests are failing because of your additions in this PR. Just run Travis again on 10.0 branch, and if it's green, then your module is breaking correct inheritance and it's reflected in any of the modules tests.

**Review by rafaelbn — APPROVED** [review]
Functionally tested in runbot 👍 
But travis should be fixed

**Comment by MiquelRForgeFlow** [comment]
@pedrobaeza The account_invoice_refund_link tests are not failing by this branch. The Travis on 10.0 branch is failing already. https://travis-ci.org/OCA/account-invoicing/builds/234196352

**Comment by gurneyalex** [comment]
@mreficent  @rafaelbn #269 fixes Travis for the 10.0 branch. Please review  :smile_cat:

**Comment by gurneyalex** [comment]
triggering a rebuild

File: account_invoice_merge/models/account_invoice.py [hunk 3]
```diff
@@ -36,10 +24,11 @@ def _get_invoice_key_cols(self):
     @api.model
     def _get_invoice_line_key_cols(self):
         fields = [
-            'name', 'origin', 'discount', 'invoice_line_tax_id', 'price_unit',
+            'name', 'origin', 'discount', 'invoice_line_tax_ids', 'price_unit',
             'product_id', 'account_id', 'account_analytic_id',
+            'uom_id'
         ]
-        for field in ['analytics_id']:
```
Comment (adrienpeiffer) [inline account_invoice_merge/models/account_invoice.py:42]: Why do you replace 'analytics_id' instead of adding the new field in this list ?

Comment (MiquelRForgeFlow) [reply account_invoice_merge/models/account_invoice.py:42]: The 'analytics_id' field was added by <code>account_analytic_plans</code> base module. But this base module was removed in v9.

Comment (adrienpeiffer) [reply account_invoice_merge/models/account_invoice.py:42]: Thanks for the explanation @mreficent

**Comment by adrienpeiffer** [comment]
@mreficent is it make sense to squash your last commits ?

**Pull request merged**